packages/core/src/session.ts

```diff
@@ -508,12 +523,18 @@ class Session {
    * @returns {void}
    */
   _onDatabaseNameResolved (database?: string): void {
-    if (!this._databaseNameResolved) {
-      const normalizedDatabase = database ?? ''
-      this._database = normalizedDatabase
-      this._readConnectionHolder.setDatabase(normalizedDatabase)
-      this._writeConnectionHolder.setDatabase(normalizedDatabase)
-      this._databaseNameResolved = true
+    if (this._isRoutingSession) {
+      this._databaseGuess = database
+      if (!this._databaseNameResolved) {
+        if (this._homeDatabaseCallback != null) {
+          this._homeDatabaseCallback(cacheKey(this._auth, this._impersonatedUser), database)
+        }
+        const normalizedDatabase = database ?? ''
+        this._database = normalizedDatabase
+        this._readConnectionHolder.setDatabase(normalizedDatabase)
+        this._writeConnectionHolder.setDatabase(normalizedDatabase)
+        this._databaseNameResolved = true
+      }
     }
```

Comment:
```suggestion
    if (!this._isRoutingSession) {
      return
    }
    this._databaseGuess = database
    if (this._databaseNameResolved) {
      return
    }
    if (this._homeDatabaseCallback != null) {
      this._homeDatabaseCallback(cacheKey(this._auth, this._impersonatedUser), database)
    }
    const normalizedDatabase = database ?? ''
    this._database = normalizedDatabase
    this._readConnectionHolder.setDatabase(normalizedDatabase)
    this._writeConnectionHolder.setDatabase(normalizedDatabase)
    this._databaseNameResolved = true
```

Less indentation and mental nesting. However, more lines. A matter of taste for sure. Just wanted to raise the option.

Reply:
Made a smaller change to this section to improve readability.